firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -79,15 +103,28 @@ private MaybeDocument getDocument(DocumentKey key, List<MutationBatch> inBatches
     ImmutableSortedMap<DocumentKey, MaybeDocument> results = emptyMaybeDocumentMap();
 
     List<MutationBatch> batches = mutationQueue.getAllMutationBatchesAffectingDocumentKeys(keys);
-    for (DocumentKey key : keys) {
-      // TODO: PERF: Consider fetching all remote documents at once rather than
-      // one-by-one.
-      MaybeDocument maybeDoc = getDocument(key, batches);
+    List<MaybeDocument> docs = getDocumentsInternal(keys, batches);
+    for (MaybeDocument maybeDoc : docs) {
+      // TODO: Don't conflate missing / deleted.
+      if (maybeDoc == null) {
+        maybeDoc = new NoDocument(maybeDoc.getKey(), SnapshotVersion.NONE, /*hasCommittedMutations=*/ false);
+      }
+      results = results.insert(maybeDoc.getKey(), maybeDoc);
+    }
+    return results;
+  }
+
+  ImmutableSortedMap<DocumentKey, MaybeDocument> getDocuments(Map<DocumentKey, MaybeDocument> docsByKey) {
```

Comment:
This avoids accessing the local database and just applies pending batches to the given documents.